File: src/main/java/de/aquafun3d/challenge/utils/TimerService.java
package de.aquafun3d.challenge.utils;

import de.aquafun3d.challenge.Main;
import net.kyori.adventure.text.Component;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.GameMode;
import org.bukkit.entity.Player;
import org.bukkit.plugin.Plugin;

import java.io.IOException;
import java.time.LocalDateTime;

public class TimerService {

	enum TimerState {PAUSED, RUNNING}
	private volatile TimerService instance = this;
	private Boolean reversed = false;
	private TimerState state = TimerState.PAUSED;
	private int lastSec = 0;
	private int sec = 0;

	public TimerService(Plugin plugin){
		if(Main.challengeConfig.contains("time")){
			sec = Main.challengeConfig.getInt("time");
		}
		Bukkit.getScheduler().scheduleSyncRepeatingTask(plugin,() ->  {
			if(state == TimerState.RUNNING && lastSec != LocalDateTime.now().getSecond()){
				lastSec = LocalDateTime.now().getSecond();
				if (reversed) {
					if(sec <= 0){
						timeRunsOut();
					}
					sec--;
				} else {
					sec++;
				}
				if(state == TimerState.RUNNING){
					for(Player p : Bukkit.getOnlinePlayers()){
						p.sendActionBar(Component.text(ChatColor.GOLD + getTimerString()));
					}
				}else{
					for(Player p : Bukkit.getOnlinePlayers()){
						setTimerActionbar(p);
					}
				}
			}
		},5L,5L);
	}

	public void toggle(){
		if(state != TimerState.RUNNING){
			if(Main.challengeConfig.contains("time") && Main.challengeConfig.getInt("time") != 0){
				sec = Main.challengeConfig.getInt("time");
			}
			state = TimerState.RUNNING;
		}else {
			try {
				Main.challengeConfig.set("time",sec);
			} catch (IOException e) {
				e.printStackTrace();
			}
			state = TimerState.PAUSED;
		}
	}

	public void reset(){
		state = TimerState.PAUSED;
		sec = 0;
		try {
			Main.challengeConfig.set("time",0);
		} catch (IOException e) {
			e.printStackTrace();
		}
	}

	public boolean isPaused(){
		return state == TimerState.PAUSED;
	}

	public boolean isReversed(){
		return reversed;
	}

	public String getTimerString(){
		String hms;
		if(sec < 3600){
			hms = String.format("%02d:%02d", sec / 60, sec % 60);
		} else if (sec < 3600 * 24) {
			hms = String.format("%02d:%02d:%02d", sec / 60 / 60, sec / 60 % 60, sec % 60);
		} else {
			hms = String.format("%d Days, %02d:%02d:%02d", sec / 60 / 60 / 24, sec / 60 / 60, sec / 60 % 60, sec % 60);
		}
		if(reversed){
			hms = ChatColor.GOLD + "" + "" + hms;
			hms += ChatColor.GREEN + " remaining";
		}
		return hms;
	}

	private void timeRunsOut(){
		Utils.atAll(ChatColor.GOLD + "Time is up!");
		Utils.atAll(ChatColor.RED + "Challenge has stopped");
		for(Player p : Bukkit.getOnlinePlayers()){
			p.setGameMode(GameMode.SPECTATOR);
		}
		Main.timerService.getInstance().toggle();
	}

	public synchronized TimerService getInstance(){
		return instance;
	}

	public void setTimerActionbar(Player p){
		p.sendActionBar(Component.text("" + ChatColor.GREEN + "" + ChatColor.ITALIC + "Timer paused"));
	}

	public int getSec() {
		return sec;
	}

	public void setSec(int sec) {
		this.sec = sec;
	}

	public void setReversed(boolean reversed){
		this.reversed = reversed;
	}
}

File: src/main/java/com/aquafun3d/challenge/utils/TimerService.kt
package com.aquafun3d.challenge.utils

import com.aquafun3d.challenge.Main
import net.md_5.bungee.api.ChatMessageType
import net.md_5.bungee.api.chat.TextComponent
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.GameMode
import org.bukkit.entity.Player
import org.bukkit.plugin.Plugin
import java.io.IOException
import java.time.LocalDateTime

class TimerService(plugin: Plugin) {

	enum class TimerState {PAUSED, RUNNING}
	@Volatile
	private var instance: TimerService? = this
	private var reversed = false//false = +, true = -
	private var state: TimerState = TimerState.PAUSED
	private var lastSec = 0
	private var sec = 0

	//initialize the runnable timer
	init {
		Bukkit.getScheduler().scheduleSyncRepeatingTask( plugin, {
			if (state === TimerState.RUNNING && lastSec != LocalDateTime.now().second) {
				lastSec = LocalDateTime.now().second
				if (reversed) {
					if(sec <= 0){
						timeRunsOut()
					}
					sec--
				} else {
					sec++
				}
			}
			if (state === TimerState.RUNNING) {
				for (p in Bukkit.getOnlinePlayers()) {
					p.spigot().sendMessage(ChatMessageType.ACTION_BAR,TextComponent(getTimerString()))
				}
			} else {
				for (p in Bukkit.getOnlinePlayers()) {
					setTimerActionbar(p)
				}
			}
		}, 5L, 5L)
	}

	//toggle timer running
	fun toggle() {
		state = if(getState() != TimerState.RUNNING){
			val cfg = Main.timerConfig
			if (cfg!!.getInt("timer.time") != 0){
				setSec(cfg.getInt("timer.time"));
			}
			TimerState.RUNNING
		}else{
			TimerState.PAUSED
		}
	}

	//reset timer
	fun reset() {
		state = TimerState.PAUSED
		sec = 0
		val cfg = Main.timerConfig
		try {
			cfg!!["timer.time"] = 0
		} catch (e: IOException) {
			e.printStackTrace()
		}
	}

	//check if timer is paused
	fun isPaused(): Boolean {
		return state === TimerState.PAUSED
	}

	//returns formatted string of timers time
	fun getTimerString(): String {
		var hms: String
		hms = if (sec < 3600) {
			String.format("%02d:%02d", sec / 60, sec % 60)
		} else if (sec < 3600 * 24) {
			String.format("%02d:%02d:%02d", sec / 60 / 60, sec / 60 % 60, sec % 60)
		} else {
			String.format("%d Days, %02d:%02d:%02d", sec / 60 / 60 / 24, sec / 60 / 60, sec / 60 % 60, sec % 60)
		}
		if (reversed) {
			hms = ChatColor.GOLD.toString() + "" + "" + hms
			hms += ChatColor.GREEN.toString() + " remaining"
		} else {
			hms = ChatColor.GOLD.toString() + ""  + "" + hms
		}
		return hms
	}

	//Ends challenge if time runs out
	private fun timeRunsOut(){
		Settings.atAll(ChatColor.GOLD.toString() + "Time is up!")
		Settings.atAll(ChatColor.RED.toString() + "Challenge has stopped")
		Bukkit.getOnlinePlayers().forEach { p: Player? -> p!!.gameMode = GameMode.SPECTATOR }
		Main.timerService?.getInstance()?.toggle()
	}

	//check if timer is reversed
	fun isReversed(): Boolean {
		return reversed
	}

	//return second timer is running
	fun getSec(): Int {
		return this.sec
	}

	//returns state of the timer
	private fun getState(): TimerState {
		return state
	}

	//make timer reverse
	fun setReversed(reversed: Boolean) {
		this.reversed = reversed
	}

	//set timer state
	fun setState(state: TimerState?) {
		this.state = state!!
	}

	//set timer seconds
	fun setSec(sec: Int) {
		this.sec = sec
	}

	//Get timer instance
	@Synchronized
	fun getInstance(): TimerService? {
		return instance
	}

	//display actionbar for timer to player
	fun setTimerActionbar(p: Player){
		p.spigot().sendMessage(ChatMessageType.ACTION_BAR, TextComponent( "" + ChatColor.GREEN + "" + ChatColor.ITALIC + "Timer paused"))
	}

}